### ISSUE TITLE
Feature Request: More control of Notebook PDF Export

### ISSUE THREAD

**Issue by rickhg12hs** [jupyterlab/jupyterlab#5199]
It would be useful if more control was possible when exporting a notebook as PDF.  For example,

1. Optionally respecting open/closed cells for PDF generation rather than always rendering closed/collapsed cells.
2. Font selection
3. Page layout (portrait/landscape, page size, margins, etc.)
4. And many more options, I'm sure. ... Possibly for the other export formats too.

If there is an export control/config file somewhere, I wouldn't mind trying to edit it as an interim step.

**Comment by t-makaro** [comment]
The export to pdf is done with [nbconvert](https://github.com/jupyter/nbconvert). In order to change things like fonts, page layout, etc., one needs to create a nbconvert template. This entails using [jinja](http://jinja.pocoo.org/) to template LaTeX. To do this in a general manner is quite difficult. Especially for fonts. Unfortunately, I think that something this general is out of sight. (Though I suppose someone could attempt to make a jlab extension that automatically makes templates though that may have it's own complications and restrictions)

It is possible to create your own template with your style preferences and then set the export within jlab to use it. You can see an example of this over at [my repo](https://github.com/t-makaro/nb_pdf_template).

For point 1, if jlab uses meta-data to store whether a cell is collapsed then it wouldn't be to hard to implement an nbconvert preprocessor to do this. (We probably don't need a new preprocessor; it may just be a case of using an existing one.)

**Comment by t-makaro** [comment]
In terms of reports and not showing the code. Nbconvert is adding a nice [shortcut](https://github.com/jupyter/nbconvert/pull/825) for this from the command line hopefully in the next release. If you want more control over nbconvert you'll have to use the command line interface and friends. If you have more questions about using nbconvert, I'd be happy to assist as I've spent a lot of time working with nbconvert and exporting to PDFs. We could find a way to communicate outside of GitHub or ask a question over in either [my repo](https://github.com/t-makaro/nb_pdf_template) or in [nbconvert](https://github.com/jupyter/nbconvert)

**Comment by timkpaine** [comment]
Lantern adds it right under the current export menu, email uses it to email pdf of the notebook.

**Comment by rickhg12hs** [comment]
Wow!  Awesome stuff!

It'd be cool if there was a selector (or cell magic) to indicate which cells should/shouldn't be rendered.

**Comment by t-makaro** [comment]
Yes, you can absolutely add new exporters to the menu. It shouldn't be to hard to subclass the pdfexporter to remove collapsed cells be default and/or all code cells (leaving the output). Both of these should even be possible without a custom exporter. It would just require changing the Jupyter config file. We just wouldn't want to add a load of exporters of every possible combination of use cases.

Now that I think about it. It could be possible to click "export to pdf" and have it open a little window with options to enable/disable different preprocessors. That wouldn't let one use different fonts/page sizes, but it would still give a load of options.

**Comment by gmoutso** [comment]
Is this still open in light of  #5015? I am trying to acchieve the --no-input nbconvert option from the jupyterlab "export as" menu.

**Comment by gmoutso** [comment]
I opened #8629.

**Comment by timkpaine** [comment]
referencing this here, adding two "no code" exporters: https://github.com/timkpaine/jupyterlab_nbconvert_nocode

**Comment by timkpaine** [comment]
@t-makaro [jupyterlab_email](https://github.com/timkpaine/jupyterlab_email) does this, it has some preconfigured templates and it lets you pick and choose options. Its a little annoying to do and a little tricky to make extensible, but it could be possible:

**Comment by MMesch** [comment]
Seems to be a subinstance of issue #17994 that groups the need for a more elaborate export dialog.
